Peers need to check each other's dust limit

t-bast opened the pull request:
Since HTLCs below this amount will not appear in the commitment tx, they are effectively converted to miner fees. The peer could use this to grief you by broadcasting its commitment once it contains a lot of dust HTLCs.

I just stumbled upon issue #696 which mentioned that a while ago, but we didn't update the spec accordingly.
However, implementations do already enforce that.

I also added a lower bound on `dust_limit_satoshis` that ensures that transactions will correctly propagate with default relay policies.

Fixes #696 and #905

TheBlueMatt commented on 02-peer-protocol.md:
```diff
@@ -239,6 +239,7 @@ The receiving node MAY fail the channel if:
   - it considers `channel_reserve_satoshis` too large.
   - it considers `max_accepted_htlcs` too small.
   - it considers `dust_limit_satoshis` too small and plans to rely on the sending node publishing its commitment transaction in the event of a data loss (see [message-retransmission](02-peer-protocol.md#message-retransmission)).
```
Should we drop this? I believe c-lightning and lnd both don't implement this at all today, and I'm not sure if we need to?

t-bast replied:
I think it's still useful, and I kept a lower bound on eclair, because you may want to avoid letting your peer create a commit tx that you think will not easily propagate through the bitcoin network because it doesn't satisfy default policy settings.

If you think you'll rely on their commit tx with `option_dataloss_protect`, you want to be sure their commit tx can get to miners' mempools, don't you?

TheBlueMatt replied:
> If you think you'll rely on their commit tx with option_dataloss_protect, you want to be sure their commit tx can get to miners' mempools, don't you?

Yea, that's a good point. Should we clarify that a suggested lower-bound is 330 (which IIRC is the exact current network dust threshold)?

t-bast replied:
Done in https://github.com/lightningnetwork/lightning-rfc/pull/894/commits/f16ea995173c432400702c502633d906827267f1

t-bast replied:
Update: I instead added a list of the network dust threshold in https://github.com/lightningnetwork/lightning-rfc/pull/894/commits/40cfede1b7f325949f8cb565cb724fdd71cf7cfc which implementers can refer to when deciding what lower bound they want to enforce.

TheBlueMatt replied:
I don't think that's really sufficient - there needs to be guidance in what implementations should be allowed to set the dust limit to, as otherwise we end up back where we are today where some nodes arbitrarily force-closed based on dust limit and there isn't a consensus. I'd strongly prefer if we either agree to set it to the max value for a standard output, or we set it to 330 and then fix the close negotiation logic via #896.

t-bast replied:
Right, I agree with that. Both options have pros and cons:

- if we all decide to enforce a minimum `dust_limit_satoshis = 546 sat`, we don't have issues at closing and #896 isn't even needed
- otherwise if we enforce a minimum `dust_limit_satoshis = 546 sat`, this allows smaller outputs on the commit tx but we need to do something at closing time (#896 or some other solution).

Note that an alternative to #896 which would also fix the issue would be to disallow non-segwit scripts in `shutdown`. Now that segwit is widely supported, this would fix a few headaches.

I'd like feedback from other implementers on this choice, @rustyrussell @Roasbeef what do you think?

t-bast replied:
I opened an issue to discuss the `dust_limit_satoshis` _lower_ bounds in https://github.com/lightningnetwork/lightning-rfc/issues/905

I'd like to treat this in a separate PR than this one, which fixes a different issue that we should have fixed a long time ago (adding a _higher_ bound on our peer's `dust_limit_satoshis`).

t-bast commented:
There can be issues with dust limits below 546, as pointed out [here](https://github.com/ACINQ/eclair/pull/1900#issuecomment-899365359).

This is also a concern with future segwit versions, for which we don't know the bitcoin network dust threshold (when using `option_shutdown_anysegwit`). I believe we need a new mechanism to handle that kind of issues.

TheBlueMatt commented:
> There can be issues with dust limits below 546, as pointed out here.

That's a good point, I'd like to hear from lnd and c-lightning folks whether they think this matters given they currently don't enforce this at all. I presume they force-close channels if the closing transaction doesn't confirm in some time?

> This is also a concern with future segwit versions, for which we don't know the bitcoin network dust threshold (when using option_shutdown_anysegwit). I believe we need a new mechanism to handle that kind of issues.

Right, this seems to be a general problem - the only thing you can realistically do here is force-close if nothing confirms in some time because at least you know all the scripts used in that transaction.

t-bast commented:
AFAICT the only place where peers are free to use different types of scripts is during mutual close.

The spec could list in the mutual close requirements the dust threshold for each type of script (e.g. 546 for non-segwit, 330 for segwit, ??? for taproot) and both nodes should enforce these static checks and prune outputs accordingly. If your peer doesn't, this will automatically result in a force-close since the closing signatures won't match.

TheBlueMatt commented:
> ??? for taproot

I tried to ask about this on `#bitcoin-core-dev` and basically got told to fuck off and not rely on relay dust limits for anything, so I guess for now just assume it won't change cause Core doesn't change things and hope for the best.

> The spec could list in the mutual close requirements the dust threshold for each type of script (e.g. 546 sats for non-segwit, 330 sats for segwit, ??? for taproot) and both nodes should enforce these static checks and prune outputs accordingly. If one of the peers doesn't, this will automatically result in a force-close since the closing signatures won't match.

That is an incompatibility, I think - one issue with `closing_signed` today is that it talks about how nodes can unilaterally drop their own output (which would be required for this to work), but then talks about how if the `fee_satoshis` field is the same as the one it sent it's done, even though the transaction may be different. Maybe we should fix it in #847?

If we fix the spec to *actually* allow unilateral dropping of a node's output, we could reasonably just say that nodes can send a dust threshold down to 330, but have to enforce that counterparties have removed their output during closing if its lower than the scripts real dust threshold.

Crypt-iQ commented:
In case anybody was wondering, I believe p2tr outputs will have a dust limit defined by Core's `GetDustThreshold` func as 330 sats, which is the same as p2wsh outputs. It may be a good idea to place the different output type's dust values calculated at dust_relay_fee of 3000sat/kvb somewhere?

Roasbeef reviewed (approved):
LGTM ⛷

Roasbeef commented on 02-peer-protocol.md:
```diff
@@ -238,7 +238,12 @@ The receiving node MAY fail the channel if:
+  - it considers `dust_limit_satoshis` too small and plans to rely on the sending
+    node publishing its commitment transaction in the event of a data loss
+    (see [message-retransmission](02-peer-protocol.md#message-retransmission)).
+    A recommended lower bound is 330 satoshis, which matches the bitcoin network's
```
Would be good IMO to add the derivation of this value somewhere as errata. I've seen some discussion lately between developers where both sides start with the same base level knowledge, but end up deriving conflicting values for dust depending on where they apply the witness discount (IIRC).

t-bast replied:
I'll add a section in BOLT 3, I agree that many people will likely find that helpful.

Crypt-iQ replied:
I have a pull to btcd to export this info, the body of the pull shows my calculations for various output types: https://github.com/btcsuite/btcd/pull/1742

t-bast replied:
I added details about the calculation in https://github.com/lightningnetwork/lightning-rfc/pull/894/commits/40cfede1b7f325949f8cb565cb724fdd71cf7cfc

TheBlueMatt commented:
I feel like this would be a good candidate for a bLIP - I don't think we will reasonably consider this something that "most nodes should implement", but instead a totally-optional cool thing for those for whom its easy.

t-bast commented:
As discussed during yesterday's spec meeting, I added option 2 of #905 in https://github.com/lightningnetwork/lightning-rfc/pull/894/commits/3e7771e2dc486a0fc8494e712da43714a1b62c39

- we explicitly reject dust limits below 354 sats to ensure there won't be propagation issues with default relay policies
- we drop support for non-segwit scripts in shutdown
- if at closing time we realize that an output will prevent propagation, we force-close

TheBlueMatt commented:
Looks good to me. Is it worth mentioning in the rationale section that older nodes will send non-segwit shutdown scripts and you may wish to allow them for backwards-compatibility today, taking note of the force-close requirement at shutdown.

t-bast commented:
> Is it worth mentioning in the rationale section that older nodes will send non-segwit shutdown scripts and you may wish to allow them for backwards-compatibility today, taking note of the force-close requirement at shutdown.

You're right, it's hard to understand without any context. I rebased and added some rationale in https://github.com/lightningnetwork/lightning-rfc/pull/894/commits/9ff138715dea52858f6bdeb1b3d2f208a74157c5

Crypt-iQ commented on 03-transactions.md:
```diff
@@ -484,6 +486,112 @@ A node:
+A p2pkh input is at least 148 bytes:
+
+- 36 bytes for the previous output (32 bytes hash + 4 bytes index)
+- 1 byte for the script sig length
```
nit: switch this line with below so script sig length is right before script sig?

t-bast replied:
Done in https://github.com/lightningnetwork/lightning-rfc/pull/894/commits/dff76b23e81a42184e861d7d7b996a81e578f206

Crypt-iQ commented on 03-transactions.md:
```diff
@@ -484,6 +486,112 @@ A node:
+A p2wpkh input is at least 67 bytes (depending on the signature length):
+
+- 36 bytes for the previous output (32 bytes hash + 4 bytes index)
+- 1 byte for the script sig length
```
same here?

t-bast replied:
Done in https://github.com/lightningnetwork/lightning-rfc/pull/894/commits/dff76b23e81a42184e861d7d7b996a81e578f206

Crypt-iQ commented on 03-transactions.md:
```diff
@@ -484,6 +486,112 @@ A node:
+- 36 bytes for the previous output (32 bytes hash + 4 bytes index)
+- 1 byte for the script sig length
+- 4 bytes for the sequence
+- 26 bytes for the witness (with the 75% segwit discount applied):
```
could mention it's a rounding down from 26.75 vbytes -> 26 vbytes?

t-bast replied:
Done in https://github.com/lightningnetwork/lightning-rfc/pull/894/commits/dff76b23e81a42184e861d7d7b996a81e578f206

Crypt-iQ commented on 02-peer-protocol.md:
```diff
@@ -270,6 +269,7 @@ The receiving node MUST fail the channel if:
   - `funding_pubkey`, `revocation_basepoint`, `htlc_basepoint`, `payment_basepoint`, or `delayed_payment_basepoint`
 are not valid secp256k1 pubkeys in compressed format.
   - `dust_limit_satoshis` is greater than `channel_reserve_satoshis`.
+  - `dust_limit_satoshis` is smaller than `354 satoshis` (see [BOLT 3](03-transactions.md#dust-limits)).
```
I think this should only apply for `option_shutdown_anysegwit` and otherwise allow a lower bound of `330 satoshis`.

t-bast replied:
> I think this should only apply for option_shutdown_anysegwit and otherwise allow a lower bound of 330 satoshis.

I disagree, because `option_shutdown_anysegwit` can be activated *after* the channel is opened.
For example, Alice and Bob open a channel without `option_shutdown_anysegwit` and set `dust_limit` to 330 sats.
Then later they both activate `option_shutdown_anysegwit` (because why not?).
If they didn't use an `upfront_shutdown_script`, they're allowed to use a segwit v1+ script when closing the channel, which can be an issue for dust limits below 354 sats.
So let's be safe and use 354 sats which is future-proof.

Crypt-iQ replied:
I see, that makes sense.

TheBlueMatt reviewed (approved):
I didn't check all the dust limits, but I believe you :)

t-bast commented:
Merging this PR as discussed during the last spec meetings and explicitly approved since.
